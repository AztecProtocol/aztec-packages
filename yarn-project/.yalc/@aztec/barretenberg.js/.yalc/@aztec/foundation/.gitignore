dest
node_modules
.cache
.env
*.log
*.swp
.tsbuildinfo
.pnp.cjs
.pnp.loader.mjs