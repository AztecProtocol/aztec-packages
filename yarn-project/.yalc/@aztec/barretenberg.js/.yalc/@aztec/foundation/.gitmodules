[submodule "build-system"]
	path = build-system
	url = git@github.com:AztecProtocol/build-system.git
