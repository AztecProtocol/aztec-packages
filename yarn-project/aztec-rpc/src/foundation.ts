import nodeCrypto from 'crypto';
import { Keccak } from 'sha3';

export const randomBytes = (len: number) => {
  return nodeCrypto.randomBytes(len) as Buffer;
};

export function keccak256(input: Buffer | string) {
  const inputBuf = typeof input === 'string' ? Buffer.from(input) : input;
  const hash = new Keccak(256);
  return hash.update(inputBuf).digest();
}
