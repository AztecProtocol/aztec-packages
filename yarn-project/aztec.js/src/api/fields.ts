export { Point, Fr, Fq, GrumpkinScalar } from '@aztec/foundation/fields';
