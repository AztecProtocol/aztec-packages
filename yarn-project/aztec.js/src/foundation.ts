export function sleep(ms: number) {
  return new Promise(resolve => setTimeout(resolve, ms));
}

export class Timer {
  private start: number;

  constructor() {
    this.start = new Date().getTime();
  }

  public ms() {
    return new Date().getTime() - this.start;
  }

  public s() {
    return (new Date().getTime() - this.start) / 1000;
  }
}

export function* backoffGenerator() {
  const v = [1, 1, 1, 2, 4, 8, 16, 32, 64];
  let i = 0;
  while (true) {
    yield v[Math.min(i++, v.length - 1)];
  }
}

export async function retry<Result>(fn: () => Promise<Result>, name = 'Operation', backoff = backoffGenerator()) {
  while (true) {
    try {
      return await fn();
    } catch (err: any) {
      const s = backoff.next().value;
      if (s === undefined) {
        throw err;
      }
      console.log(`${name} failed. Will retry in ${s}s...`);
      console.log(err);
      await sleep(s * 1000);
      continue;
    }
  }
}

// Call `fn` repeatedly until it returns true or timeout.
// Both `interval` and `timeout` are seconds.
// Will never timeout if the value is 0.
export async function retryUntil<T>(fn: () => Promise<T | undefined>, name = '', timeout = 0, interval = 1) {
  const timer = new Timer();
  while (true) {
    const result = await fn();
    if (result) {
      return result;
    }

    await sleep(interval * 1000);

    if (timeout && timer.s() > timeout) {
      throw new Error(name ? `Timeout awaiting ${name}` : 'Timeout');
    }
  }
}
