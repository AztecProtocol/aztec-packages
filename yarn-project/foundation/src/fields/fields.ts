import { AztecAddress } from '../aztec-address/index.js';
import { toBigIntBE, toBufferBE, toHex } from '../bigint-buffer/index.js';
import { randomBytes } from '../crypto/index.js';
import { BufferReader } from '../serialize/buffer_reader.js';

/**
 * Fr represents a field of integers modulo the prime number MODULUS.
 * It provides utility functions to work with elements in this field, such as conversions between different representations and checks for equality and zero values. The elements can be serialized to and deserialized from byte buffers or strings.
 * Some use cases include working with cryptographic operations and finite fields.
 */
export class Fr {
  static ZERO = new Fr(0n);
  static MODULUS = 0x30644e72e131a029b85045b68181585d2833e84879b9709143e1f593f0000001n;
  static MAX_VALUE = Fr.MODULUS - 1n;
  static SIZE_IN_BYTES = 32;

  /**
   * The numeric value of the field element as a bigint.
   */
  public readonly value: bigint;

  constructor(value: boolean | bigint | number | Fr | AztecAddress) {
    const isFr = (value: boolean | bigint | number | Fr | AztecAddress): value is Fr | AztecAddress =>
      !!(value as Fr).toBigInt;
    this.value = isFr(value) ? value.toBigInt() : BigInt(value);
    if (this.value > Fr.MAX_VALUE) {
      throw new Error(`Fr out of range ${value}.`);
    }
  }

  /**
   * Generates a random Fr or Fq instance with a value modulo the respective class' MODULUS.
   * This method uses randomBytes to generate a random 32-byte buffer, converts it to a bigint
   * and takes the modulus of the result with the class' MODULUS constant.
   *
   * @returns A new Fr or Fq instance with a random value.
   */
  static random() {
    const r = toBigIntBE(randomBytes(Fr.SIZE_IN_BYTES)) % Fr.MODULUS;
    return new Fr(r);
  }

  /**
   * Returns a new zero-value field.
   * @returns A new zero-value field.
   */
  static zero() {
    return new Fr(0n);
  }

  /**
   * Create an instance of the corresponding class (Fr or Fq) from a Buffer or a BufferReader.
   * Reads 'SIZE_IN_BYTES' bytes from the given Buffer or BufferReader and constructs an instance with the decoded value.
   * If the input is a Buffer, it is internally converted to a BufferReader before reading.
   * Throws an error if the input length is invalid or the decoded value is out of range.
   *
   * @param buffer - The Buffer or BufferReader containing the bytes representing the value.
   * @returns An instance of the corresponding class (Fr or Fq) with the decoded value.
   */
  static fromBuffer(buffer: Buffer | BufferReader) {
    const reader = BufferReader.asReader(buffer);
    return new Fr(toBigIntBE(reader.readBytes(Fr.SIZE_IN_BYTES)));
  }

  /**
   * Create a Fr instance from a hex-encoded string.
   * The input 'address' can be either prefixed with '0x' or not, and should have exactly 64 hex characters.
   * Throws an error if the input length is invalid or the address value is out of range.
   *
   * @param address - The hex-encoded string representing the field element.
   * @returns A Fr instance.
   */
  static fromString(address: string) {
    return Fr.fromBuffer(Buffer.from(address.replace(/^0x/i, ''), 'hex'));
  }

  /**
   * Converts the value of the instance to a buffer with a specified length.
   * The method uses the provided value and size in bytes to create a buffer representation
   * of the numeric value. This can be useful for serialization and communication purposes.
   *
   * @returns A buffer representing the instance's value.
   */
  toBuffer() {
    return toBufferBE(this.value, Fr.SIZE_IN_BYTES);
  }

  /**
   * Converts the value of the Fr or Fq class instance to a hexadecimal string.
   * The resulting string is prefixed with '0x' and represents the bigint value
   * in base 16.
   *
   * @param padTo32 - Whether to pad the string to 32 bytes.
   * @returns A hex-encoded string representing the value of the class instance.
   */
  toString(padTo32 = false): `0x${string}` {
    return toHex(this.value, padTo32);
  }

  /**
   * Retrieves the underlying bigint.
   * This method mostly exists to match user expectations, as value is already public.
   * @returns The underlying bigint.
   */
  public toBigInt(): bigint {
    return this.value;
  }

  /**
   * Returns a shortened string representation of the Fr value, formatted with '0x' prefix and ellipsis in the middle.
   * The resulting string has first 10 characters (including '0x') and last 4 characters of the full hexadecimal value.
   *
   * @returns A shorter, human-readable string representation of the Fr value.
   */
  toShortString() {
    const str = this.toString();
    return `${str.slice(0, 10)}...${str.slice(-4)}`;
  }

  /**
   * Checks if the provided Fr instance is equal to the current instance.
   * Two instances are considered equal if their 'value' properties are the same.
   *
   * @param rhs - The Fr instance to compare with the current instance.
   * @returns A boolean indicating whether the two instances are equal.
   */
  equals(rhs: Fr) {
    return this.value === rhs.value;
  }

  /**
   * Check if the instance value is zero.
   * The method returns true if the value of the instance is 0n (zero in BigInt representation),
   * otherwise, it returns false.
   *
   * @returns A boolean indicating whether the instance value is zero or not.
   */
  isZero() {
    return this.value === 0n;
  }

  /**
   * Converts the current value of the Fq or Fr instance to a friendly JSON representation.
   * The output will be a hexadecimal string prefixed with '0x'.
   *
   * @returns A '0x' prefixed hexadecimal string representing the current value.
   */
  toFriendlyJSON() {
    return this.toString();
  }
}

/**
 * Fq represents a field element in a prime finite field with modulus defined by the constant MODULUS.
 * It provides methods for creating, manipulating, and comparing field elements, as well as converting
 * them to/from different data types like Buffers and hex strings. Field elements are used in various
 * cryptographic protocols and operations, such as elliptic curve cryptography.
 *
 * @example
 * const fqElem = new Fq(BigInt("123456789"));
 * const randomFqElem = Fq.random();
 * const fromBufferFqElem = Fq.fromBuffer(buffer);
 */
export class Fq {
  static MODULUS = 0x30644e72e131a029b85045b68181585d97816a916871ca8d3c208c16d87cfd47n;
  static MAX_VALUE = Fr.MODULUS - 1n;
  static SIZE_IN_BYTES = 32;

  constructor(
    /**
     * The element's value as a bigint in the finite field.
     */
    public readonly value: bigint,
  ) {
    if (value > Fq.MAX_VALUE) {
      throw new Error(`Fq out of range ${value}.`);
    }
  }

  /**
   * Generates a random Fr or Fq instance with a value within the range of their respective modulus.
   * The random value is generated from a byte array of length equal to SIZE_IN_BYTES, then truncated
   * to the appropriate modulus before creating the new Fr or Fq instance.
   *
   * @returns A new Fr or Fq instance with a randomly generated value.
   */
  static random() {
    const r = toBigIntBE(randomBytes(64)) % Fq.MODULUS;
    return new this(r);
  }

  /**
   * Create an instance of the calling class (Fr or Fq) from a given buffer or BufferReader.
   * Reads SIZE_IN_BYTES from the provided buffer and converts it to a bigint, then creates a new instance
   * with that value. Throws an error if the value is out of range for the calling class.
   *
   * @param buffer - The input buffer or BufferReader containing the bytes representing the value.
   * @returns An instance of the calling class (Fr or Fq) initialized with the bigint value.
   */
  static fromBuffer(buffer: Buffer | BufferReader) {
    const reader = BufferReader.asReader(buffer);
    return new this(toBigIntBE(reader.readBytes(this.SIZE_IN_BYTES)));
  }

  /**
   * Converts the bigint value of the instance to a Buffer representation.
   * The output buffer has a fixed size, determined by the 'SIZE_IN_BYTES' constant.
   *
   * @returns A Buffer containing the byte representation of the instance's value.
   */
  toBuffer() {
    return toBufferBE(this.value, Fq.SIZE_IN_BYTES);
  }

  /**
   * Converts the Fq value to a hexadecimal string representation.
   * The resulting string is prefixed with '0x' and contains the exact number of hex characters required
   * to represent the numeric value of this instance.
   *
   * @returns A hexadecimal string representing the Fq value.
   */
  toString() {
    return toHex(this.value);
  }

  /**
   * Check if the value of the current instance is zero.
   * This function compares the internal 'value' property with 0n (BigInt representation of zero).
   * Returns true if the value is zero, otherwise returns false.
   *
   * @returns A boolean indicating whether the value is zero or not.
   */
  isZero() {
    return this.value === 0n;
  }

  /**
   * Converts the value of the Fr or Fq instance to a friendly JSON format.
   * The output is a hexadecimal string representation of the value with '0x' prefix.
   *
   * @returns A string representing the value in the JSON format.
   */
  toFriendlyJSON() {
    return this.toString();
  }
}
