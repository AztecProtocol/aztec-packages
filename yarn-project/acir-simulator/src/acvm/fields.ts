import { AztecAddress, EthAddress, Fr } from '@aztec/foundation';

export type ACVMField = `0x${string}`;
export type ACVMWitness = Map<number, ACVMField>;

export const ZERO_ACVM_FIELD: ACVMField = `0x${Buffer.alloc(32).toString('hex')}`;
export const ONE_ACVM_FIELD: ACVMField = `0x${'00'.repeat(31)}01`;

function adaptBufferSize(originalBuf: Buffer) {
  const buffer = Buffer.alloc(32);
  if (originalBuf.length > buffer.length) {
    throw new Error('Buffer does not fit in 32 bytes');
  }
  originalBuf.copy(buffer, buffer.length - originalBuf.length);
  return buffer;
}

export function toACVMField(value: AztecAddress | EthAddress | Fr | Buffer | boolean | number): `0x${string}` {
  if (typeof value === 'boolean') {
    return value ? ONE_ACVM_FIELD : ZERO_ACVM_FIELD;
  }

  let buffer;

  if (Buffer.isBuffer(value)) {
    buffer = value;
  } else if (typeof value === 'number') {
    buffer = Buffer.alloc(32);
    buffer.writeUInt32BE(value, 28);
  } else {
    buffer = value.toBuffer();
  }

  return `0x${adaptBufferSize(buffer).toString('hex')}`;
}

export function fromACVMField(field: `0x${string}`): Fr {
  const buffer = Buffer.from(field.slice(2), 'hex');
  return Fr.fromBuffer(buffer);
}
