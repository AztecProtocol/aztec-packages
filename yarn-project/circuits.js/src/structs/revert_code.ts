import { Fr } from '@aztec/foundation/fields';
import { BufferReader, FieldReader } from '@aztec/foundation/serialize';

import { inspect } from 'util';

enum RevertCodeEnum {
  OK = 0,
  APP_LOGIC_REVERTED = 1,
  TEARDOWN_REVERTED = 2,
  BOTH_REVERTED = 3,
}

function isRevertCodeEnum(value: number): value is RevertCodeEnum {
  return (
    value === RevertCodeEnum.OK ||
    value === RevertCodeEnum.APP_LOGIC_REVERTED ||
    value === RevertCodeEnum.TEARDOWN_REVERTED ||
    value === RevertCodeEnum.BOTH_REVERTED
  );
}

/**
 * Wrapper class over a field to safely represent a revert code.
 */
export class RevertCode {
  private code: number;
  private constructor(e: RevertCodeEnum) {
    this.code = e.valueOf();
  }
  static readonly OK: RevertCode = new RevertCode(RevertCodeEnum.OK);
  static readonly APP_LOGIC_REVERTED: RevertCode = new RevertCode(RevertCodeEnum.APP_LOGIC_REVERTED);
  static readonly TEARDOWN_REVERTED: RevertCode = new RevertCode(RevertCodeEnum.TEARDOWN_REVERTED);
  static readonly BOTH_REVERTED: RevertCode = new RevertCode(RevertCodeEnum.BOTH_REVERTED);

  public equals(other: RevertCode): boolean {
    return this.code === other.code;
  }

  public isOK(): boolean {
    return this.equals(RevertCode.OK);
  }

  public getDescription() {
    switch (this.code) {
      case RevertCodeEnum.OK:
        return 'OK';
      case RevertCodeEnum.APP_LOGIC_REVERTED:
        return 'Application logic reverted';
      case RevertCodeEnum.TEARDOWN_REVERTED:
        return 'Teardown reverted';
      case RevertCodeEnum.BOTH_REVERTED:
        return 'Both reverted';
      default:
        return `Unknown RevertCode: ${this.code}`;
    }
  }

  /**
   * Having different serialization methods allows for
   * decoupling the serialization for producing the content commitment hash
   * (where we use fields)
   * from serialization for transmitting the data.
   */

  private static readonly PREIMAGE_SIZE_IN_BYTES = 32;
  public toHashPreimage(): Buffer {
    const padding = Buffer.alloc(RevertCode.PREIMAGE_SIZE_IN_BYTES - RevertCode.PACKED_SIZE_IN_BYTES);
    return Buffer.concat([padding, this.toBuffer()]);
  }

  private static readonly PACKED_SIZE_IN_BYTES = 1;
  public toBuffer(): Buffer {
    const b = Buffer.alloc(RevertCode.PACKED_SIZE_IN_BYTES);
    b.writeUInt8(this.code, 0);
    return b;
  }

  public toField(): Fr {
    return new Fr(this.toBuffer());
  }

  public getSerializedLength(): number {
    return this.toBuffer().length;
  }

  public static fromField(fr: Fr): RevertCode {
    if (!isRevertCodeEnum(fr.toNumber())) {
      throw new Error(`Invalid RevertCode: ${fr.toNumber()}`);
    }
    return new RevertCode(fr.toNumber());
  }

  public static fromFields(fields: Fr[] | FieldReader): RevertCode {
    const reader = FieldReader.asReader(fields);
    return RevertCode.fromField(reader.readField());
  }

  public static fromBuffer(buffer: Buffer | BufferReader): RevertCode {
    const reader = BufferReader.asReader(buffer);
    const code = reader.readBytes(RevertCode.PACKED_SIZE_IN_BYTES).readUInt8(0);
    if (!isRevertCodeEnum(code)) {
      throw new Error(`Invalid RevertCode: ${code}`);
    }
    return new RevertCode(code);
  }

  private static readonly NUM_OPTIONS = 4;
  static random(): RevertCode {
    return new RevertCode(Math.floor(Math.random() * RevertCode.NUM_OPTIONS));
  }

  [inspect.custom]() {
    return `RevertCode<${this.code.toString()}>`;
  }
}
