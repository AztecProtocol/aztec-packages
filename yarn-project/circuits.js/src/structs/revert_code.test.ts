import { Fr } from '@aztec/foundation/fields';

import { RevertCode } from './revert_code.js';

describe('revert_code', () => {
  it.each([RevertCode.OK, RevertCode.APP_LOGIC_REVERTED, RevertCode.TEARDOWN_REVERTED, RevertCode.BOTH_REVERTED])(
    'should serialize properly',
    revertCode => {
      expect(revertCode.getSerializedLength()).toBe(1);

      const hashPreimage = revertCode.toHashPreimage();
      expect(hashPreimage).toMatchSnapshot();
      expect(hashPreimage.length).toBe(32);

      const buf = revertCode.toBuffer();
      expect(buf).toMatchSnapshot();
      expect(RevertCode.fromBuffer(buf)).toEqual(revertCode);

      const field = revertCode.toField();
      expect(field).toMatchSnapshot();
      expect(RevertCode.fromField(field)).toEqual(revertCode);
      expect(RevertCode.fromFields([field])).toEqual(revertCode);
    },
  );

  it('should throw when deserializing from invalid buffer', () => {
    expect(() => RevertCode.fromBuffer(Buffer.from([42]))).toThrow();
    expect(() => RevertCode.fromField(new Fr(42))).toThrow();
  });
});
