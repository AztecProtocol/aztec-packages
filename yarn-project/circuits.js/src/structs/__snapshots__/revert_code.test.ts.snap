// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`revert_code should serialize properly 1`] = `
{
  "data": [
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
  ],
  "type": "Buffer",
}
`;

exports[`revert_code should serialize properly 2`] = `
{
  "data": [
    0,
  ],
  "type": "Buffer",
}
`;

exports[`revert_code should serialize properly 3`] = `
{
  "type": "Fr",
  "value": "0x0000000000000000000000000000000000000000000000000000000000000000",
}
`;

exports[`revert_code should serialize properly 4`] = `
{
  "data": [
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    1,
  ],
  "type": "Buffer",
}
`;

exports[`revert_code should serialize properly 5`] = `
{
  "data": [
    1,
  ],
  "type": "Buffer",
}
`;

exports[`revert_code should serialize properly 6`] = `
{
  "type": "Fr",
  "value": "0x0000000000000000000000000000000000000000000000000000000000000001",
}
`;

exports[`revert_code should serialize properly 7`] = `
{
  "data": [
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    2,
  ],
  "type": "Buffer",
}
`;

exports[`revert_code should serialize properly 8`] = `
{
  "data": [
    2,
  ],
  "type": "Buffer",
}
`;

exports[`revert_code should serialize properly 9`] = `
{
  "type": "Fr",
  "value": "0x0000000000000000000000000000000000000000000000000000000000000002",
}
`;

exports[`revert_code should serialize properly 10`] = `
{
  "data": [
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    3,
  ],
  "type": "Buffer",
}
`;

exports[`revert_code should serialize properly 11`] = `
{
  "data": [
    3,
  ],
  "type": "Buffer",
}
`;

exports[`revert_code should serialize properly 12`] = `
{
  "type": "Fr",
  "value": "0x0000000000000000000000000000000000000000000000000000000000000003",
}
`;
